### PR TITLE
CircleCI refactor

### PR DESCRIPTION
**Pull request by stephenroller**
**Patch description**
For a while, we've been using the legacy 2.0 syntax of CircleCI, rather than the newer 2.1 syntax. For a while, this was because we used a feature that was only available in 2.0 (dynamically running jobs), but now we run our full CI on every commit.

This PR refactors us to the newer 2.1 syntax. This also allows us to fix a small bug, where data caches were being indexed on unit tests, rather than teachers, and causing invalid caches more frequently than necessary.

**Testing steps**
CI

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -239,459 +333,146 @@ jobs:
...
-      - save_cache:
-          key: nvidia-downloads-20200516
-          paths:
-            - "~/nvidia-downloads/"
-      - restore_cache:
-          key: deps-20200921-gpu14-{{ checksum "requirements.txt" }}
-      - <<: *setup
-      - <<: *installtorchgpu14
-      - <<: *installdeps
-      - <<: *installdetectrondeps
```
Comment (EricMichaelSmith) [inline .circleci/config.yml:357]: Hmm how do things like `installdetectrondeps` get run now if they're not explicitly called?

File: .circleci/config.yml [hunk 2]
```diff
@@ -239,459 +333,146 @@ jobs:
...
+          more_installs:
+            - installtorchgpu16
+          install_cuda: true
+          cachename: nightly
+          marker: nightly_gpu
+          pytest_flags: -v -s
 
   crowdsourcing_tests:
-    <<: *standard_cpu37
+    executor: small_cpu37
```
Comment (EricMichaelSmith) [inline .circleci/config.yml:428]: hmm - the change is just to conserve resources? If so, that's fine with me - you have a better sense of how much each size CPU can handle

File: tests/test_quickstart.sh [hunk 3]
```diff
@@ -1,48 +0,0 @@
-#!/bin/bash
-# Copyright (c) Facebook, Inc. and its affiliates.
-# This source code is licensed under the MIT license found in the
-# LICENSE file in the root directory of this source tree.
-
-set -e -x  # error and exit on any failure; print the commands being run
-
-# test installation & help commands
```
Comment (EricMichaelSmith) [inline tests/test_quickstart.sh:8]: Hmm I guess this was redundant?

**Review by EricMichaelSmith — APPROVED** [review]
Seems like a good idea! I had a few questions

File: .circleci/config.yml (hunk 2 above)
Comment (stephenroller) [reply .circleci/config.yml:428]: Yeah, just lowering our bill

File: .circleci/config.yml (hunk 1 above)
Comment (stephenroller) [reply .circleci/config.yml:357]: This was missed, it needs to be added back in

File: tests/test_quickstart.sh (hunk 3 above)
Comment (stephenroller) [reply tests/test_quickstart.sh:8]: I threw it away!

**Pull request merged**